service/storage_service.cc

```diff
@@ -4905,6 +4905,20 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                     // we already did read barrier above
                     result.status = raft_topology_cmd_result::command_status::success;
                 break;
+                case raft_topology_cmd::command::barrier_and_drain: {
+                    const auto version = _topology_state_machine._topology.version;
+                    co_await container().invoke_on_all([version] (storage_service& ss) -> future<> {
+                        const auto current_version = ss._shared_token_metadata.get()->get_version();
+                        slogger.debug("Got raft_topology_cmd::barrier_and_drain, version {}, current version {}",
+                            version, current_version);
+                        if (current_version == version) {
+                            co_await ss._shared_token_metadata.stale_versions_in_use();
```

Comment:
> Why aren't we waiting for stale versions if current_version == version? Couldn't this violate the post-condition of barrier_and_drain which says that there are no stale requests?

The condition `current_version == version` can be violated only if the topology change coordinator changed (moved to another node) during `barrier_and_drain` and managed to increment version concurrently. In this case the current `barrier_and_drain` is moot and we can skip all the waiting. `stale_versions_in_use()` waits for all the the previous versions, this will be done in the `barrier_and_drain` RPC from the new coordinator.

Reply:
So we can skip the waiting, because there shouldn't be anyone waiting for the result, if the system works as intended. But isn't it better to fail in this case? This is also correct, and more in line with the intent of the barrier. If our assumption don't hold for some reason, we will learn about it quickly rather than masking the problem with wrong behavior.

Reply:
> So we can skip the waiting, because there shouldn't be anyone waiting for the result, if the system works as intended. 

Yes.

> If our assumption don't hold for some reason, we will learn about it quickly rather than masking the problem with wrong behavior.

Makes sense, ok, will update.